pkg/agent/cluster.go

```diff
@@ -0,0 +1,330 @@
+			czero.installHistory.ClusterKubeAPISeen = true
+		}
+
+		configmap, _ := czero.API.Kube.IsBootstrapConfigMapComplete()
+		if configmap {
+			logrus.Info("bootstrap configMap status is complete")
+			return true, nil
+		}
+	}
+
+	if agentRestAPILive {
+		// First time we see the agent Rest API
+		if !czero.installHistory.RestAPISeen {
+			logrus.Info("node zero Agent API Initialized")
+			czero.installHistory.RestAPISeen = true
+		}
+		logrus.Trace("getting cluster info from Node Zero Agent API")
+		clusterState, _ := czero.Get()
+
+		// TODO(lranjbar)[AGENT-172]: Add CheckHostValidations
+		// if !czero.installHistory.AgentHostValidationsPassed {
+		// 	validations := CheckHostValidations()
+		// 	if validations {
+		// 		czero.installHistory.AgentHostValidationsPassed = true
+		// 	}
+		// }
+
+		czero.PrintInstallStatus(clusterState)
+		czero.installHistory.RestAPICurrentClusterStatus = *clusterState.Status
+
+		// Update Install History object when we see these states
+		czero.updateInstallHistoryClusterStatus(clusterState)
+
+		stopped, _ := czero.HasStoppedInstalling(*clusterState.Status)
+		if stopped {
+			logrus.Error("cluster has stopped installing")
+			errored, _ := czero.HasErrored(*clusterState.Status)
+			if errored {
+				return false, errors.New("cluster installation has stopped due to errors")
+			}
+			return false, errors.New("cluster has stopped installing and/or insallation was cancelled")
+		}
+
+	}
+
+	// both API's are not available
+	if !agentRestAPILive && !clusterKubeAPILive {
+		logrus.Debug("current API Status: Node Zero Agent API: down, Cluster Kube API: down")
+		if !czero.installHistory.RestAPISeen && !czero.installHistory.ClusterKubeAPISeen {
+			logrus.Debug("nero zero Agent API never initialized. Cluster API never initialized")
+			logrus.Warn("unable to detect installation. Cluster install has either not initalized or was not started")
+			return false, nil
+		}
+
+		if czero.installHistory.RestAPISeen && !czero.installHistory.ClusterKubeAPISeen {
+			logrus.Debug("cluster API never initialized")
+			logrus.Debugf("cluster install status last seen was: %s", czero.installHistory.RestAPICurrentClusterStatus)
+			return false, errors.New("cluster installation did not complete")
+		}
+	}
+
+	logrus.Debug("bootstrap is not complete. sleeping... ")
+	return false, nil
+}
+
+// IsInstallComplete Determine if the cluster has completed installation.
+func (czero *Cluster) IsInstallComplete() (bool, error) {
+	return true, nil
+}
+
+// IsInstalling Determine if the cluster is still installing using the models from the Agent Rest API.
+func (czero *Cluster) IsInstalling(status string) (bool, string) {
+	clusterInstallingStates := map[string]bool{
+		models.ClusterStatusAddingHosts:                 true,
+		models.ClusterStatusCancelled:                   false,
+		models.ClusterStatusInstalling:                  true,
+		models.ClusterStatusInstallingPendingUserAction: false,
```

Comment:
Q: why this status is not considered an installing one? Couldn't be, for example, that the user has to fix something to let the installation process continue?

Reply:
Yes. This state happens when the customer has to interact with the machines to push the install forward but the installation has not reached an error that would stop the install completely. 

A common scenario is that they are installing on an old BIOS machine. The mechanism we use to reboot a machine isn't supported by BIOS but if the customer intervenes and reboots the BIOS machine the installation will resume.

Reply:
I missed the previous update, it seems a better choice to let always the loop naturally reach the timeout (in case of issue), following a similar behavior of the standard wait-for